### PR TITLE
Notification implementation

### LINKED ISSUE #45: Implement Notification API
See https://w3c.github.io/wot-discovery/#exploration-directory-api-notification

### PR DESCRIPTION
**Pull request by farshidtz**
Closes #45

### CODE REVIEW

File: notification/notification.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package notification
+
+import (
+	"github.com/linksmart/thing-directory/catalog"
+)
+
+type Event struct {
+	Id   string                   `json:"id"`
```
Comment (farshidtz) [inline notification/notification.go:8]: Should be ID

File: notification/mem_storage.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	events   []Event
+	latestID uint64
+}
+
+func NewMemStorage() *MemStorage {
+	return &MemStorage{latestID: rand.Uint64()}
+}
+
+func (m MemStorage) add(event Event) error {
+	m.events = append(m.events, event)
```
Comment (farshidtz) [inline notification/mem_storage.go:18]: This should have a limited capacity to avoid memory exhaustion.

File: notification/http.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+
+	"github.com/linksmart/thing-directory/common"
+)
+
+type HTTPAPI struct {
```
Comment (farshidtz) [inline notification/http.go:11]: Calling it SSE is more appropriate (also the file name). SSE uses HTTP internally but a typical HTTP client will not be able to consumer it.

File: notification/eventType.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package notification
```
Comment (farshidtz) [inline notification/eventType.go:1]: filename should not be camel-cased. This file could me merged with notification.go.

File: notification/mem_storage.go (hunk 2 above)
Comment (dschowta) [reply notification/mem_storage.go:18]: This is just a dummy code and shall be improved

File: common/errors.go [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2014-2016 Fraunhofer Institute for Applied Information Technology FIT
 
-package catalog
+package common
```
Comment (farshidtz) [inline common/errors.go:3]: I don't think this refactoring was necessary. At list not as part of this PR.

**Review by farshidtz — COMMENTED** [review]
Added several inline comments.

File: main.go [hunk 6]
```diff
@@ -216,6 +236,9 @@ func setupHTTPRouter(config *HTTPConfig, api *catalog.HTTPAPI) (*negroni.Negroni
 	// TD validation
 	r.get("/validation", commonHandlers.ThenFunc(api.GetValidation))
 
+	//TD notification
+	r.get("/events", commonHandlers.ThenFunc(notifAPI.SubscribeEvent))
```
Comment (farshidtz) [inline main.go:240]: Have to check how will this work with oauth2 tokens.

File: notification/controller.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+			c.activeClients[s.client] = s
+			log.Printf("New subscription. %d active clients", len(c.activeClients))
+
+			// Send the missed events
+			if s.lastEventID != "" {
+				missedEvents, err := c.s.getAllAfter(s.lastEventID)
+				if err != nil {
+					log.Printf("error getting the events after ID %s: %s", s.lastEventID, err)
+				}
+				for _, event := range missedEvents {
```
Comment (farshidtz) [inline notification/controller.go:151]: `missedEvents` will be empty if err is not nil? Better put this in an `else` statement.

File: notification/ldbstorage.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	ldbstorage := &LevelDBStorage{db: db, capacity: capacity}
+	ldbstorage.latestID, err = ldbstorage.fetchLatestID()
+	if err != nil {
+		return nil, fmt.Errorf("error fetching the latest ID: %w", err)
+	}
+	return ldbstorage, nil
+}
+func (s *LevelDBStorage) add(event Event) error {
```
Comment (farshidtz) [inline notification/ldbstorage.go:45]: this function does more than it should, considering its name.

File: notification/ldbstorage.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+func (s *LevelDBStorage) addRotate(event Event) error {
+	s.wg.Add(1)
+	defer s.wg.Done()
+
+	// add new data
+	bytes, err := json.Marshal(event)
+	if err != nil {
+		return fmt.Errorf("error marshaling event: %w", err)
```
Comment (farshidtz) [inline notification/ldbstorage.go:56]: typo. change to marshalling

File: notification/sse.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	QueryParamFull    = "full"
+	HeaderLastEventID = "Last-Event-ID"
+)
+
+type SSEAPI struct {
+	controller  NotificationController
+	contentType string
+}
+
+func NewHTTPAPI(controller NotificationController, version string) *SSEAPI {
```
Comment (farshidtz) [inline notification/sse.go:24]: NewSSEAPI

File: notification/sse.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	w.Header().Set("Content-Type", a.contentType)
+
+	messageChan := make(chan Event)
+
+	lastEventID := req.Header.Get(HeaderLastEventID)
+	a.controller.subscribe(messageChan, eventTypes, full, lastEventID)
+
+	go func() {
+		<-req.Context().Done()
+		a.controller.unsubscribe(messageChan)
```
Comment (farshidtz) [inline notification/sse.go:55]: Does this close `messageChan` so that the following loop exits?

File: common/errors.go [hunk 12]
```diff
@@ -65,12 +65,12 @@ func ProblemDetailsResponse(w http.ResponseWriter, pd wot.ProblemDetails) {
 	}
 	b, err := json.Marshal(pd)
 	if err != nil {
-		log.Printf("ERROR serializing error object: %s", err)
+		log.Printf("ERROR serializing error object: %S", err)
```
Comment (farshidtz) [inline common/errors.go:68]: Please revert

File: common/errors.go [hunk 13]
```diff
@@ -65,12 +65,12 @@ func ProblemDetailsResponse(w http.ResponseWriter, pd wot.ProblemDetails) {
...
 	}
 	w.Header().Set("Content-Type", "application/problem+json")
 	w.WriteHeader(pd.Status)
 	_, err = w.Write(b)
 	if err != nil {
-		log.Printf("ERROR writing HTTP response: %s", err)
+		log.Printf("ERROR writing HTTP response: %S", err)
```
Comment (farshidtz) [inline common/errors.go:74]: Please revert

File: notification/event_type.go [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package notification
```
Comment (farshidtz) [inline notification/event_type.go:1]: Only some parts are exported. Can move these to `wot` package, under wot/discovery.go.

File: notification/sse.go (hunk 11 above)
Comment (dschowta) [reply notification/sse.go:56]: good catch. Now it is closed in the handler function of controller

File: notification/ldbstorage.go (hunk 8 above)
Comment (dschowta) [reply notification/ldbstorage.go]: changed the function name.  Structure/interface is renamed to EventQueue to match with the functionalities

File: main.go [hunk 15]
```diff
@@ -104,10 +105,29 @@ func main() {
...
-	nRouter, err := setupHTTPRouter(&config.HTTP, api)
+	// Start notification
+	var notificationStorage notification.EventQueue
+	switch config.Storage.Type {
+	case catalog.BackendLevelDB:
+		notificationStorage, err = notification.NewLevelDBEventQueue(config.Storage.DSN+"/sse", nil, 1000)
+		if err != nil {
+			panic("Failed to start LevelDB storage for SSE events:" + err.Error())
+		}
+		defer storage.Close()
```
Comment (farshidtz) [inline main.go:116]: notificationStorage.close()

**Pull request merged**